torch/autograd/functional.py

```diff
@@ -0,0 +1,717 @@
+                               "a single Tensor with a single element.")
+
+    grad_res = _autograd_grad(outputs, inputs, v, create_graph=create_graph)
+
+    vjp = _fill_in_zeros(grad_res, inputs, strict, create_graph, "back")
+
+    # Cleanup objects and return them to the user
+    outputs = _grad_postprocess(outputs, create_graph)
+    vjp = _grad_postprocess(vjp, create_graph)
+
+    return _tuple_postprocess(outputs, is_outputs_tuple), _tuple_postprocess(vjp, is_inputs_tuple)
+
+
+def jvp(func, inputs, v=None, create_graph=False, strict=False):
+    r"""Function that computes the dot product between  the Jacobian of
+    the given function at the point given by the inputs and a vector ``v``.
+
+    Args:
+        func (function): a Python function that takes Tensor inputs and returns
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function ``func``.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the input of ``func``. This argument is optional when
+            ``func``'s input contains a single element and (if it is not provided) will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict is ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect that there exists an input
+            such that all the outputs are independent of it. If ``False``, we return zeros as the jvp for
+            said inputs, which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
+        result (tuple of Tensor or Tensor): result of the dot product with the same shape
+            as the output.
+
+    Example::
+
+        >>> def exp_reducer(x):
+        ...   return x.exp().sum(dim=1)
+        >>> inputs = torch.rand(4, 4)
+        >>> v = torch.ones(4, 4)
+        >>> jvp(exp_reducer, inputs, v)
+        (tensor([6.3090, 4.6742, 7.9114, 8.2106]),
+         tensor([6.3090, 4.6742, 7.9114, 8.2106]))
+
+        >>> jvp(exp_reducer, inputs, v, create_graph=True)
+        (tensor([6.3090, 4.6742, 7.9114, 8.2106], grad_fn=<SumBackward1>),
+         tensor([6.3090, 4.6742, 7.9114, 8.2106], grad_fn=<SqueezeBackward1>))
+
+        >>> def adder(x, y):
+        ...   return 2 * x + 3 * y
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> v = (torch.ones(2), torch.ones(2))
+        >>> jvp(adder, inputs, v)
+        (tensor([2.2399, 2.5005]),
+         tensor([5., 5.]))
+
+    Note::
+
+        The jvp is currently computed using the double backward trick as we don't have support for
```

Comment:
is there a reference you can use for the double backward trick?

Reply:
I wasn't sure what was the original reference for it. I think this is widely known for people working on autograd.

Reply:
Yes, but more people than just autograd experts will read this documentation.  I think it's fine to just describe it briefly e.g. "compute the backwards of the backward (sometimes called the double backwards trick)"